src/Castle.Core/DynamicProxy/Generators/MethodSignatureComparer.cs

```diff
@@ -120,28 +121,39 @@ private bool EqualSignatureTypes(Type x, Type y, MethodInfo xm = null, MethodInf
 					return false;
 				}
 
-				for (var i = 0; i < xArgs.Length; ++i)
+				if (IsCovariantReturnTypes(x, y, xm, ym) == false)
 				{
-					 if(!EqualSignatureTypes(xArgs[i], yArgs[i])) return false;
+					for (var i = 0; i < xArgs.Length; ++i)
+					{
+						 if(!EqualSignatureTypes(xArgs[i], yArgs[i])) return false;
```

Comment:
`IsCovariantReturnTypes` gets called for every generic parameter. That'll obviously do nothing as the `MethodInfo`s are null, probably nothing that could be improved without it being a premature micro-optimisation though.

Reply:
It just ocurred to me that the obvious thing to do is to move the covariant return type check from `EqualSignatureTypes` to `EqualReturnTypes`. Then it will be performed _only_ for return types, and we only need it in one location (instead of in three).